java/src/org/openqa/selenium/grid/node/config/NodeOptions.java

```diff
@@ -344,143 +344,124 @@ private SessionFactory createSessionFactory(String clazz, Capabilities stereotyp
+            }
+
+            // if config lacks 'stereotype' setting
+            if (!configMap.containsKey("stereotype")) {
+              throw new ConfigException("Found config with no 'stereotype' setting! " + configMap);
+            }
+
+            // add config to list
+            configList.add(configMap);
+            // prepare for next config
+            configMap = new HashMap<>();
+          } else {
+            // add setting to config
+            configMap.put(values[0], unquote(values[1]));
+          }
+        }
 
-              List<WebDriverInfo> infos = new ArrayList<>();
-              ServiceLoader.load(WebDriverInfo.class).forEach(infos::add);
+        // if no configs were found
+        if (configList.isEmpty()) {
+          throw new ConfigException("No driver configs were found!");
+        }
 
-              driversMap.forEach(
-                  configMap -> {
-                    if (!configMap.containsKey("stereotype")) {
-                      throw new ConfigException(
-                          "Driver config is missing stereotype value. " + configMap);
-                    }
+        List<DriverService.Builder<?, ?>> builderList = new ArrayList<>();
+        ServiceLoader.load(DriverService.Builder.class).forEach(builderList::add);
+
+        List<WebDriverInfo> infoList = new ArrayList<>();
+        ServiceLoader.load(WebDriverInfo.class).forEach(infoList::add);
+
+        // iterate over driver configs
+        configList.forEach(thisConfig -> {
+          // create Capabilities object from stereotype of this config
+          Capabilities confStereotype = JSON.toType(thisConfig.get("stereotype"), Capabilities.class);
+
+          // extract driver executable path from this config
+          String webDriverExecutablePath = thisConfig.get("webdriver-executable");
+          // if executable path is specified
+          if (null != webDriverExecutablePath) {
+            // create File object from executable path string
+            File webDriverExecutable = new File(webDriverExecutablePath);
+            // if specified path isn't a file
+            if (!webDriverExecutable.isFile()) {
+              LOG.warning("Driver executable does not seem to be a file! " + webDriverExecutablePath);
+            }
+
+            // if specified path isn't executable
+            if (!webDriverExecutable.canExecute()) {
+              LOG.warning("Driver file exists but does not seem to be a executable! " + webDriverExecutablePath);
+            }
+
+            // add specified driver executable path to capabilities
+            confStereotype = new PersistentCapabilities(confStereotype)
+              .setCapability("se:webDriverExecutable", webDriverExecutablePath);
+          }
+
+          Capabilities stereotype = enhanceStereotype(confStereotype);
+          String configName = thisConfig.getOrDefault("display-name", "Custom Slot Config");
```

Comment:
I'm pretty sure that there was no way to get to this point in the prior implementation without `display-name` being defined. The new implementation explicitly checks for the presence of this setting and bails out if it's absent. Was there a plan for a "custom slot config" feature? If so, I can remove the check for the setting and allow for this possibility.

Reply:
If we determine that there's a use case for "custom slot config", adding support for this should be easy.